---
title: "R_assignment"
output: 
  html_document: 
    toc: yes
---
######instruction#####


maize files with missing data marked as "?" are named as Chr_*_zea_genotype_without_replacement.txt

maize files with missing data marked as "-" are named as Chr_*_zea_genotype_after_replacement.txt

teosinte files with missing data marked as "?" are named as Chr_*_teosinte_genotype_without_replacement.txt

teosinte files with missing data marked as "-" are named as Chr_*_teosinte_genotype_after_replacement.txt

Note: graphic figure will be generated at the working directory that you set, please check if they are there if you can not find them when directly running them in rmd file (some of them are big).

######Data Inspection#####

set working directiory and liabrary (Pls change it to your own directory before running any code)
```{r}
setwd("~/Desktop/homework/R_assignment")
library(gtools)
library(ggplot2)
library('dplyr')
library(reshape2)
```
read the files
```{r}
snp_p<-read.table("snp_position.txt",sep="\t", header=T,stringsAsFactors = F)
genotypic_data<-read.table("fang_et_al_genotypes.txt",header=T,stringsAsFactors = F)
```

Inspect data structure
```{r,echo=TRUE}
	str(snp_p)
	str(genotypic_data)
```
For SNP information there are 15 columns and 983 SNPs

For genotypic data, there are 2782 lines with 983 SNPs, for each row, the first three columns are "Sample_ID" "JG_OTU" "Group", then followed by 983 SNPs

We also noticed that the Position column of snp and chromosme information file is "chr", this indicates there are text in that column
```{r}
	table(snp_p$Position)
  table(snp_p$Chromosome)
```
We got:

	  ... 95897171  96216125  96216463  98507715  multiple   unknown 
    ...    1         1         1         1        11        27
There are 11 rows with multiple position and 27 rows with unknown position

Also
         9    multiple    unknown 
          60        6       27 
There are 6 row has multiple chromosme position and 27 has unknown chromosome position
Then inspect how many different genotype do we have:
```{r}
	table(genotypic_data$Group)
```	
Then we got:

	TRIPS ZDIPL ZLUXR ZMHUE ZMMIL ZMMLR ZMMMR ZMPBA ZMPIL ZMPJA ZMXCH ZMXCP ZMXIL ZMXNO ZMXNT ZPERR 
   	22    15    17    10   290  1256    27   900    41    34    75    69     6     7     4     9 

We have three maize groups: ZMMIL, ZMMLR, and ZMMMR total number is 290+1256+27=1573
And three teosinte groups: ZMPBA, ZMPIL, and ZMPJA total number is 900+41+34=975

######Data transformation#####

fist get the maize and teosinte genotype
```{r}
zea<-subset(genotypic_data,Group=="ZMMIL"|Group=="ZMMLR"|Group=="ZMMMR")
teosinte<-subset(genotypic_data,Group=="ZMPBA"|Group=="ZMPIL"|Group=="ZMPJA")
```

then transpose with 1st column of original file as header and header of origianal file as row name
then only keep the rows with genotypic information
```{r}
teosinte_t= setNames(data.frame(t(teosinte[,-1])), teosinte[,1])
zea_t= setNames(data.frame(t(zea[,-1])), zea[,1])
teosinte_t<-teosinte_t[c(-1,-2),]
zea_t<-zea_t[c(-1,-2),]
```
get the SNP_ID, Chrosome and Position fron snp_p
```{r}
snp_temp<-snp_p[,c(1,3,4)]
```

combin the transpose file and snp_id to generate genotypic files for teosinte and zea
```{r}
zea_geno<-cbind(snp_temp,zea_t)
teosinte_geno<-cbind(snp_temp,teosinte_t)
```
replace the ? with -
```{r}
zea_geno_replaced<-data.frame(lapply(zea_geno, function(x) {gsub("?", "-", x,fixed=T)}))
teosinte_geno_replaced<-data.frame(lapply(teosinte_geno, function(x) {gsub("?", "-", x,fixed=T)}))
```

Generate the files:
For Zea
First is 10 files with ? marked as missing data
```{r}
for(i in 1:10){
  k<-NULL
  k2<-NULL
  k<-subset(zea_geno, Chromosome==i)
  k$Position<-as.character(k$Position)
  k2<-k[mixedorder(k$Position),]
  write.table(k2,file = paste0("Chr_",i,"_zea_genotype_without_replecment.txt"),sep = "\t",row.names = F,quote = F)
}
```
10 files with "-" marked as missing data
```{r}
for(i in 1:10){
  k<-NULL
  k2<-NULL
  k<-subset(zea_geno_replaced, Chromosome==i)
  k$Position<-as.character(k$Position)
  k2<-k[mixedorder(k$Position),]
  write.table(k2,file = paste0("Chr_",i,"_zea_genotype_after_replecment.txt"),sep = "\t",row.names = F,quote = F)
}
```
For teosinte
First is 10 files with ? marked as missing data
```{r}
for(i in 1:10){
  k<-NULL
  k2<-NULL
  k<-subset(teosinte_geno, Chromosome==i)
  k$Position<-as.character(k$Position)
  k2<-k[mixedorder(k$Position),]
  write.table(k2,file = paste0("Chr_",i,"_teosinte_genotype_without_replecment.txt"),sep = "\t",row.names = F,quote = F)
}
```
10 files with "-" marked as missing data
```{r}
for(i in 1:10){
  k<-NULL
  k2<-NULL
  k<-subset(teosinte_geno_replaced, Chromosome==i)
  k$Position<-as.character(k$Position)
  k2<-k[mixedorder(k$Position),]
  write.table(k2,file = paste0("Chr_",i,"_teosinte_genotype_after_replecment.txt"),sep = "\t",row.names = F,quote = F)
}
```

#Plotting
```{r}
P<-as.data.frame(table(snp_temp$Chromosome))
P$Var1<-as.character(P$Var1)
P<-P[mixedorder(P$Var1),]
P$Freq<-as.numeric(P$Freq)
names(P)<-c("Chr","Number.of.SNPs")
P$Chr <- factor(P$Chr, levels=unique(P$Chr))
png(height=900, width=500, pointsize=30, file="SNP_number_by_chromosome.png")
ggplot(P,aes(x=Chr,y=Number.of.SNPs))+geom_point()+xlab("Chromosome")+ylab("Number of SNPs")

```
#Homozygosity

First we need transfrom the data into: 1 As Heterozygous 2 As Homozygous NA as Missing
```{r}
for (i in 1:nrow(genotypic_data)){
  
  for (j in 4:ncol(genotypic_data)) {
    if(genotypic_data[i,j]=="A/A"|genotypic_data[i,j]=="T/T"|genotypic_data[i,j]=="G/G"|genotypic_data[i,j]=="C/C"){
      genotypic_data[i,j]=2 
    }else{
      if(genotypic_data[i,j]=="?/?"){
        genotypic_data[i,j]=c("NA")
      }else{
        genotypic_data[i,j]=1
      }
    }
  }
}

```

Then transpose the data to calculate frequency of Homozygous, heterozygous and NA of each sample and adjust these value back to the genotypic folder with first three columns

```{r}
genotyp_info<-genotypic_data[,c(1:3)]
genotyp_info$homozygous<-NA
genotyp_info$heterozygous<-NA
genotyp_info$Missing<-NA
temp= setNames(data.frame(t(genotypic_data[,-1])), genotypic_data[,1])
i<-NULL
j<-NULL
for(i in 1:nrow(genotyp_info)){
  m<-table(temp[,i])
  s<-m[1]+m[2]+m[3]
  #get heterozygous freq
  genotyp_info[i,5]<-m[1]/s
  #get homozygous freq
  genotyp_info[i,4]<-m[2]/s
  #get missing freq
  genotyp_info[i,6]<-m[3]/s
}
```
sorting the data

```{r}
genotyp_info<-genotyp_info[mixedorder(genotyp_info$Sample_ID),]
genotyp_info<-genotyp_info[mixedorder(genotyp_info$Group),]
```

First draw the frequency map of all samples
```{r}
temp1<-genotyp_info[,c(1,4:6)]
d <- melt(temp1, id.vars="Sample_ID")
png(height=1000, width=500, file="Proportion_of_Allele_frequency.png")
ggplot(d, aes(Sample_ID,value, fill=variable)) + 
  geom_bar(stat = "identity")+xlab("Sample_ID")+ylab("frequency")
```

Then draw the frequency map using the max value of each allele types within a group

```{r}
temp2<-genotyp_info[,c(3:6)]
d2 <- melt(temp2, id.vars="Group")
png(height=1000, width=500, file="Proportion_of_Allele_frequency_by_group.png")
ggplot(d2, aes(x=Group,y=value, fill=variable)) + 
  geom_bar(stat = "identity",position=position_dodge())+xlab("Sample_ID")+ylab("MAx Frequency")

```
done

We can also check the proportion of missing data of each genotype:
```{r}
png(height=1000, width=500,  file="Missing_genotypic_data_rate.png")
ggplot(genotyp_info,aes(Sample_ID, Missing))+
  geom_bar(stat = "identity")+xlab("Sample_ID")+ylab("Proportion of Missing Data")
```

Hmmm...some genotypes might need to be removed from the dataset....